src/Functions/array/arrayLevenshtein.cpp

```diff
@@ -0,0 +1,350 @@
+
+        Columns holders(num_arguments);
+        std::vector<const ColumnArray *> columns(num_arguments);
+
+        for (size_t i = 0; i < num_arguments; ++i)
+        {
+            holders[i] = arguments[i].column->convertToFullColumnIfConst();
+            if (holders[i]->size() != input_rows_count)
+                throw Exception(
+                    ErrorCodes::LOGICAL_ERROR,
+                    "Function {} has unequal number of rows in columns: "
+                    "expected {}, got {} for column {}",
+                    getName(),
+                    input_rows_count,
+                    holders[i]->size(),
+                    holders[i]->getName());
+            columns[i] = assert_cast<const ColumnArray*>(holders[i].get());
+        }
+        switch (T::impl)
+        {
+            case 0:
+                return simpleLevenshteinImpl(columns);
+            case 1:
+                return weightedLevenshteinImpl(columns, false);
+            case 2:
+                return weightedLevenshteinImpl(columns, true);
+        }
+    }
+private:
+    ColumnPtr simpleLevenshteinImpl(std::vector<const ColumnArray *> columns) const
+    {
+        const ColumnArray * column_lhs = columns[0];
+        const ColumnArray * column_rhs = columns[1];
+        auto res = ColumnUInt32::create();
+        ColumnUInt32::Container & res_values = res->getData();
+        res_values.resize(column_lhs->size());
+        for (size_t row = 0; row < column_lhs->size(); row++)
+        {
+            // Effective Levenshtein realization from Common/levenshteinDistance
+            Array lhs = (*column_lhs)[row].safeGet<Array>();
+            Array rhs = (*column_rhs)[row].safeGet<Array>();
+            const size_t m = lhs.size();
+            const size_t n = rhs.size();
+            if (m==0 || n==0)
+            {
+                res_values[row] = static_cast<UInt32>(std::max(m, n));
+                continue;
+            }
+            PODArrayWithStackMemory<size_t, 32> v0(n + 1);
+
+            iota(v0.data() + 1, n, size_t(1));
+
+            for (size_t j = 1; j <= m; ++j)
+            {
+                v0[0] = j;
+                size_t prev = j - 1;
+                for (size_t i = 1; i <= n; ++i)
+                {
+                    size_t old = v0[i];
+                    v0[i] = std::min(prev + (lhs[j - 1] != rhs[i - 1]),
+                                     std::min(v0[i - 1], v0[i]) + 1);
+                    prev = old;
+                }
+            }
+            res_values[row] = static_cast<UInt32>(v0[n]);
+        }
+        return res;
+    }
+
+    ColumnPtr weightedLevenshteinImpl(std::vector<const ColumnArray *> columns, bool similarity) const
+    {
+        for (size_t i = 0; i < 2; i++)
+        {
+            const ColumnArray * hs_column = columns[i];
+            const ColumnArray * weights_column = columns[i+2];
+            for (size_t row = 0; row < hs_column->size(); row++)
```

Comment:
Does it mean, that all rows must be checked in the beginning of the execution? So, the block will start proceeding only after it's ready?

Maybe, the better place for it is inside the main cycle over the rows below?

Reply:
I thought of it tonight, and it looks safe. The whole block should be there already, otherwise `size_t input_rows_count` wouldn't make any sense.